File: qc-core/src/main/java/com/eager/questioncloud/core/domain/point/ChargePointPaymentRepositoryImpl.java
package com.eager.questioncloud.core.domain.point;

import static com.eager.questioncloud.core.domain.point.QChargePointPaymentEntity.chargePointPaymentEntity;

import com.eager.questioncloud.core.common.PagingInformation;
import com.eager.questioncloud.exception.CustomException;
import com.eager.questioncloud.exception.Error;
import com.querydsl.jpa.impl.JPAQueryFactory;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class ChargePointPaymentRepositoryImpl implements ChargePointPaymentRepository {
    private final ChargePointPaymentJpaRepository chargePointPaymentJpaRepository;
    private final JPAQueryFactory jpaQueryFactory;

    @Override
    public ChargePointPayment save(ChargePointPayment chargePointPayment) {
        return chargePointPaymentJpaRepository.save(ChargePointPaymentEntity.from(chargePointPayment)).toModel();
    }

    @Override
    public Boolean isCompletedPayment(Long userId, String paymentId) {
        return jpaQueryFactory.select(chargePointPaymentEntity.paymentId)
            .from(chargePointPaymentEntity)
            .where(
                chargePointPaymentEntity.paymentId.eq(paymentId),
                chargePointPaymentEntity.userId.eq(userId),
                chargePointPaymentEntity.chargePointPaymentStatus.eq(ChargePointPaymentStatus.PAID))
            .fetchFirst() != null;
    }

    @Override
    public ChargePointPayment getChargePointPaymentForApprove(String paymentId) {
        ChargePointPaymentEntity resultEntity =
            jpaQueryFactory.select(chargePointPaymentEntity)
                .from(chargePointPaymentEntity)
                .where(
                    chargePointPaymentEntity.paymentId.eq(paymentId),
                    chargePointPaymentEntity.chargePointPaymentStatus.eq(ChargePointPaymentStatus.ORDERED))
                .fetchFirst();

        if (resultEntity == null) {
            throw new CustomException(Error.NOT_FOUND);
        }

        return resultEntity.toModel();
    }

    @Override
    public ChargePointPayment findByPaymentId(String paymentId) {
        return chargePointPaymentJpaRepository.findByPaymentId(paymentId)
            .orElseThrow(() -> new CustomException(Error.NOT_FOUND))
            .toModel();
    }

    @Override
    public Boolean existsByPaymentId(String paymentId) {
        return jpaQueryFactory.select(chargePointPaymentEntity.paymentId)
            .from(chargePointPaymentEntity)
            .where(chargePointPaymentEntity.paymentId.eq(paymentId))
            .fetchFirst() != null;
    }

    @Override
    public List<ChargePointPayment> getChargePointPayments(Long userId, PagingInformation pagingInformation) {
        return jpaQueryFactory.select(chargePointPaymentEntity)
            .from(chargePointPaymentEntity)
            .where(chargePointPaymentEntity.userId.eq(userId))
            .offset(pagingInformation.getOffset())
            .limit(pagingInformation.getSize())
            .orderBy(chargePointPaymentEntity.id.desc())
            .fetch()
            .stream()
            .map(ChargePointPaymentEntity::toModel)
            .collect(Collectors.toList());
    }

    @Override
    public int countByUserId(Long userId) {
        return jpaQueryFactory.select(chargePointPaymentEntity.id.count())
            .from(chargePointPaymentEntity)
            .where(chargePointPaymentEntity.userId.eq(userId))
            .fetchFirst()
            .intValue();
    }
}

File: qc-domain/qc-point/qc-point-core/src/main/kotlin/com/eager/questioncloud/point/infrastructure/repository/ChargePointPaymentRepositoryImpl.kt
package com.eager.questioncloud.point.infrastructure.repository

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.common.pagination.PagingInformation
import com.eager.questioncloud.point.domain.ChargePointPayment
import com.eager.questioncloud.point.enums.ChargePointPaymentStatus
import com.eager.questioncloud.point.infrastructure.entity.ChargePointPaymentEntity
import com.eager.questioncloud.point.infrastructure.entity.QChargePointPaymentEntity.chargePointPaymentEntity
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import java.util.stream.Collectors

@Repository
class ChargePointPaymentRepositoryImpl(
    private val chargePointPaymentJpaRepository: ChargePointPaymentJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
) : ChargePointPaymentRepository {
    override fun save(chargePointPayment: ChargePointPayment): ChargePointPayment {
        return chargePointPaymentJpaRepository.save(
            ChargePointPaymentEntity.createNewEntity(
                chargePointPayment
            )
        )
            .toModel()
    }
    
    override fun update(chargePointPayment: ChargePointPayment): ChargePointPayment {
        return chargePointPaymentJpaRepository.save(
            ChargePointPaymentEntity.fromExisting(
                chargePointPayment
            )
        ).toModel()
    }
    
    override fun isCompletedPayment(userId: Long, orderId: String): Boolean {
        return jpaQueryFactory.select(chargePointPaymentEntity.orderId)
            .from(chargePointPaymentEntity)
            .where(
                chargePointPaymentEntity.orderId.eq(orderId),
                chargePointPaymentEntity.userId.eq(userId),
                chargePointPaymentEntity.chargePointPaymentStatus.eq(ChargePointPaymentStatus.CHARGED)
            )
            .fetchFirst() != null
    }
    
    override fun findByOrderIdWithLock(orderId: String): ChargePointPayment {
        return chargePointPaymentJpaRepository.findByOrderIdWithLock(orderId)
            .orElseThrow { CoreException(Error.NOT_FOUND) }
            .toModel()
    }
    
    override fun findByOrderId(orderId: String): ChargePointPayment {
        return chargePointPaymentJpaRepository.findById(orderId)
            .orElseThrow { CoreException(Error.NOT_FOUND) }
            .toModel()
    }
    
    override fun getChargePointPayments(
        userId: Long,
        pagingInformation: PagingInformation
    ): List<ChargePointPayment> {
        return jpaQueryFactory.select(chargePointPaymentEntity)
            .from(chargePointPaymentEntity)
            .where(chargePointPaymentEntity.userId.eq(userId))
            .offset(pagingInformation.offset.toLong())
            .limit(pagingInformation.size.toLong())
            .orderBy(chargePointPaymentEntity.paymentId.desc())
            .fetch()
            .stream()
            .map { entity: ChargePointPaymentEntity -> entity.toModel() }
            .collect(Collectors.toList())
    }
    
    override fun countByUserId(userId: Long): Int {
        return jpaQueryFactory.select(chargePointPaymentEntity.paymentId.count())
            .from(chargePointPaymentEntity)
            .where(chargePointPaymentEntity.userId.eq(userId))
            .fetchFirst()?.toInt() ?: 0
    }
    
    override fun deleteAllInBatch() {
        chargePointPaymentJpaRepository.deleteAllInBatch()
    }
}
